pkg/rhcos/ami.go

```diff
@@ -2,86 +2,22 @@ package rhcos
 
 import (
 	"context"
-	"time"
 
-	"github.com/aws/aws-sdk-go/aws"
-	"github.com/aws/aws-sdk-go/aws/session"
-	"github.com/aws/aws-sdk-go/service/ec2"
 	"github.com/pkg/errors"
 )
 
-const (
-	// DefaultChannel is the default RHCOS channel for the cluster.
-	DefaultChannel = "tested"
-)
-
-// AMI calculates a Red Hat CoreOS AMI.
-func AMI(ctx context.Context, channel, region string) (ami string, err error) {
-	if channel != DefaultChannel {
-		return "", errors.Errorf("channel %q is not yet supported", channel)
-	}
-
-	ssn := session.Must(session.NewSessionWithOptions(session.Options{
-		SharedConfigState: session.SharedConfigEnable,
-		Config: aws.Config{
-			Region: aws.String(region),
-		},
-	}))
-
-	svc := ec2.New(ssn)
-
-	result, err := svc.DescribeImagesWithContext(ctx, &ec2.DescribeImagesInput{
-		Filters: []*ec2.Filter{
-			{
-				Name:   aws.String("name"),
-				Values: aws.StringSlice([]string{"redhat-coreos-*"}),
-			},
-			{
-				Name:   aws.String("architecture"),
-				Values: aws.StringSlice([]string{"x86_64"}),
-			},
-			{
-				Name:   aws.String("virtualization-type"),
-				Values: aws.StringSlice([]string{"hvm"}),
-			},
-			{
-				Name:   aws.String("image-type"),
-				Values: aws.StringSlice([]string{"machine"}),
-			},
-			{
-				Name:   aws.String("owner-id"),
-				Values: aws.StringSlice([]string{"531415883065"}),
-			},
-			{
-				Name:   aws.String("state"),
-				Values: aws.StringSlice([]string{"available"}),
-			},
-		},
-	})
+// AMI fetches the HVM AMI ID of the latest Red Hat CoreOS release.
+func AMI(ctx context.Context, channel, region string) (string, error) {
+	meta, err := fetchLatestMetadata(ctx, channel)
 	if err != nil {
-		return "", errors.Wrap(err, "failed to describe AMIs")
+		return "", errors.Wrap(err, "failed to fetch RHCOS metadata")
 	}
 
-	var image *ec2.Image
-	var created time.Time
-	for _, nextImage := range result.Images {
-		if nextImage.ImageId == nil || nextImage.CreationDate == nil {
-			continue
+	for _, ami := range meta.AMIs {
+		if ami.Name == region {
```

Comment:
Do we want to land this before [the AMIs are being pushed to other regions][1]?

```console
BUILD=$(curl -sL https://rhcos-release-browser-coreos.int.open.paas.redhat.com/storage/releases/maipo/builds.json | jq -r '.builds[0]')
$ echo "${BUILD}"
47.38
$ curl -sL "https://rhcos-release-browser-coreos.int.open.paas.redhat.com/storage/releases/maipo/${BUILD}/meta.json" | jq .amis
[
  {
    "name": "us-east-1",
    "hvm": "ami-03afbcc8b72560ac8"
  }
]
```

I'd rather stick to scraping AWS until we get that, because scraping AWS has a chance at working in other regions if folks can copy the AMI over to the target region.  That would allow us to use [the other regions we're trying to use for `openshift-dev`][2].

[1]: https://github.com/openshift/installer/pull/572#issuecomment-434395142
[2]: https://github.com/openshift/installer/pull/572#issuecomment-434404697

Reply:
We can add an environment variable for the AMI. We already have the override for the QEMU image.

Reply:
> We can add an environment variable for the AMI. We already have the override for the QEMU image.

I'm ok with that, although see [this comment][1] about not having AMI/image knobs and assuming that there's a default AMI/image that is (1) available and (2) healthy enough to pivot.

[1]: https://github.com/openshift/installer/issues/281#issuecomment-423199497

Reply:
We're close to landing the multi-region again although this is going to slow our pipeline notably.  There are two things we need to do; first on the RHCOS side, implement tags, and only upload to all the regions for tags (e.g. once a day at most?).

Second though we need to do https://github.com/openshift/os/issues/289 or possibly teach the installer to take a set of override RPMs and apply them via Ignition?  Then I think we can slow the cadence of "promoted" builds, and anyone who wants to test bleeding edge OS work can use libvirt.